src/logging/logging.cpp

```diff
@@ -71,57 +82,36 @@ k4a_result_t logger_register_message_callback(k4a_logging_message_cb_t *message_
 {
+    logger_global_context_t *g_context = logger_global_context_t_get();
     k4a_result_t result = K4A_RESULT_SUCCEEDED;
 
+    // Validate parameters
     if (message_cb)
     {
-        result = K4A_RESULT_FROM_BOOL(min_level >= K4A_LOG_LEVEL_CRITICAL && min_level <= K4A_LOG_LEVEL_OFF);
-        if (K4A_SUCCEEDED(result))
-        {
-            long count_of_registered_callbacks = INC_REF_VAR(g_user_logger_cb_info_ref);
-            if (count_of_registered_callbacks == 1)
-            {
-                // We won the right to register and create logger_cb_t
-                g_user_logger_cb_info = new logger_user_cb_info_t();
-                g_user_logger_cb_info->callback = message_cb;
-                g_user_logger_cb_info->callback_context = message_cb_context;
-                g_user_log_level = min_level;
-            }
-            else if (g_user_logger_cb_info->callback == message_cb)
-            {
-                // User is calling to update the min_level
-                g_user_log_level = min_level;
-                DEC_REF_VAR(g_user_logger_cb_info_ref);
-            }
-            else
-            {
-                // Disallow the caller from registering a 2nd callback. A new callback can be established by clearing
-                // the existing callback function.
-                result = K4A_RESULT_FROM_BOOL(count_of_registered_callbacks == 1);
-                DEC_REF_VAR(g_user_logger_cb_info_ref);
-                result = K4A_RESULT_FAILED;
-            }
-        }
+        RETURN_VALUE_IF_ARG(K4A_RESULT_FAILED, min_level < K4A_LOG_LEVEL_CRITICAL);
+        RETURN_VALUE_IF_ARG(K4A_RESULT_FAILED, min_level > K4A_LOG_LEVEL_OFF);
     }
-    else if (g_user_logger_cb_info)
-    {
-        volatile logger_user_cb_info_t *logger_info = g_user_logger_cb_info;
-        g_user_logger_cb_info = nullptr;
-        DEC_REF_VAR(g_user_logger_cb_info_ref);
-
-        // NOTE: this loop could loop forever if user calls start and a new g_user_logger_cb_info is allocated, which
-        // would add a ref to g_user_logger_cb_info_ref. We don't expect there to be races between 1 thread trying to
-        // release the callback function and another thread to add one. To protect against this we would need a more
-        // elaborate locking scheme and the logging functions would having to use it, which would impact performance.
-        while (g_user_logger_cb_info_ref != 0)
-        {
-            // Wait for the parallel calls to logger_log to end
-            ThreadAPI_Sleep(10);
-        }
 
-        delete logger_info;
-        g_user_log_level = K4A_LOG_LEVEL_OFF;
+    rwlock_acquire_write(&g_context->lock);
+
+    // The user may set or clear the callback, or set the callback to the
+    // same value it was previously. It is a failure to change the callback
+    // from an existing registration.
+    if (g_context->user_callback == NULL || message_cb == NULL || g_context->user_callback == message_cb)
```

Comment:
if (```(```g_context->user_callback == NULL ```&&``` message_cb ```!=```NULL```)``` || g_context->user_callback == message_cb)

Reply:
I think this logic is correct. You allow setting the callback (g_context->user_callback == NULL) as well as clearing the callback (message_cb !=NULL).